lib/membrane_rtmp_plugin/rtmp/source/amf3/parser.ex

```diff
@@ -0,0 +1,216 @@
+
+    if (number &&& 0x01) == 1 do
+      {{:date, :previous}, rest}
+    else
+      <<date::float-size(64), rest::binary>> = rest
+
+      {{:date, date}, rest}
+    end
+  end
+
+  # array
+  defp parse_value(<<0x09, rest::binary>>) do
+    {size, rest} = parse_integer(rest)
+
+    if (size &&& 0x01) == 1 do
+      dense_array_size = size >>> 1
+
+      {assoc_array, rest} = parse_assoc_array(rest, [])
+      {dense_array, rest} = parse_dense_array(rest, dense_array_size, [])
+
+      {assoc_array ++ dense_array, rest}
+    else
+      raise "Unsupported array reference"
+    end
+  end
+
+  # object
+  defp parse_value(<<0x0A, _rest::binary>>) do
+    raise "Unsupported object type"
+  end
+
+  # xml
+  defp parse_value(<<0x0B, rest::binary>>) do
+    {size, rest} = parse_integer(rest)
+
+    # if the bit is set then we are dealing with string literal
+    if (size &&& 0x01) == 1 do
+      size = size >>> 1
+
+      <<string::binary-size(size), rest::binary>> = rest
+
+      {{:xml, string}, rest}
+    else
+      raise "Unsupported xml reference"
+    end
+  end
+
+  # byte array
+  defp parse_value(<<0x0C, rest::binary>>) do
+    {size, rest} = parse_integer(rest)
+
+    if (size &&& 0x01) == 1 do
+      size = size >>> 1
+
+      <<bytes::binary-size(size), rest::binary>> = rest
+
+      {bytes, rest}
+    else
+      raise "Unsupported byte array reference"
+    end
+  end
+
+  # vector int
+  defp parse_value(<<0x0D, _rest::binary>>) do
+    raise "Unsupported vector int type"
+  end
+
+  # vector uint
+  defp parse_value(<<0x0E, _rest::binary>>) do
+    raise "Unsupported vector uint type"
+  end
+
+  # vector double
+  defp parse_value(<<0x0F, _rest::binary>>) do
+    raise "Unsupported vector double type"
+  end
+
+  # vector object
+  defp parse_value(<<0x10, _rest::binary>>) do
+    raise "Unsupported vector object type"
+  end
+
+  # dictionary
+  defp parse_value(<<0x11, _rest::binary>>) do
+    raise "Unsupported dictionary type"
+  end
+
+  defp parse_integer(<<0::1, value::7, rest::binary>>), do: {value, rest}
+
+  defp parse_integer(<<1::1, first::7, 0::1, second::7, rest::binary>>) do
+    {bsl(first, 7) + second, rest}
+  end
+
+  defp parse_integer(<<1::1, first::7, 1::1, second::7, 0::1, third::7, rest::binary>>) do
+    {bsl(first, 14) + bsl(second, 7) + third, rest}
+  end
+
+  defp parse_integer(<<1::1, first::7, 1::1, second::7, 0::1, third::7, fourth::8, rest::binary>>) do
+    {bsl(first, 22) + bsl(second, 15) + bsl(third, 7) + fourth, rest}
+  end
+
+  defp parse_string(payload) do
+    {size, rest} = parse_integer(payload)
+
+    # if the bit is set then we are dealing with string literal
+    if (size &&& 0x01) == 1 do
+      size = size >>> 1
+
+      <<string::binary-size(size), rest::binary>> = rest
+
+      {string, rest}
+    else
+      raise "Unsupported string reference"
+    end
+  end
+
+  defp parse_assoc_array(<<0x01, rest::binary>>, acc), do: {Enum.reverse(acc), rest}
```

Comment:
Correct me if I'm wrong, but shouldn't we terminate parsing assoc array on an empty string? So the check should include `<<0x01, 0::28>>`

Reply:
As we checked, the empty UTF-8 is just a `0x01` marker